File: coding/src/com/algos/core/li18_string/revised/L125_ValidPalindrome.java
package  com.algos.core.li18_string.revised;

public class L125_ValidPalindrome {

	public static void main(String[] args) {
		L125_ValidPalindrome obj = new L125_ValidPalindrome();
		String s = "A man, a plan, a canal: Panama";
		//String s = "race a car";
		//String s = "0P";
		System.out.println(obj.isPalindrome(s));
	}
	
    // Problem: https://leetcode.com/problems/valid-palindrome/
    // Idea: Convert string to lowercase. Move front and back ptrs. Skip non-alphanumeric.
	public boolean isPalindrome(String s) {
        s = s.toLowerCase();
        int l=0, r=s.length()-1;
        while(l<r){
            if(!(s.charAt(l) >= 'a' && s.charAt(l) <= 'z') &&
              !(s.charAt(l) >= '0' && s.charAt(l) <= '9')){
                l++;
                continue;
            }else if(!(s.charAt(r) >= 'a' && s.charAt(r) <= 'z') &&
                    !(s.charAt(r) >= '0' && s.charAt(r) <= '9')){
                r--;
                continue;
            }else{
                if(s.charAt(l) != s.charAt(r)){
                    return false;
                }
                l++; r--;
            }
        }
        return true;
    }
}

File: dsa/concepts/18_strings/python/l125_valid_palindrome.py
# Problem: https://leetcode.com/problems/valid-palindrome/
# Idea: Convert string to lowercase. Move left and right ptrs. Skip
# non-alphanumeric.
def is_palindrome(s):
    s = s.lower()
    l = 0
    r = len(s) - 1

    while l < r:
        if not ('a' <= s[l] <= 'z' or '0' <= s[l] <= '9'):
            l += 1
            continue

        if not ('a' <= s[r] <= 'z' or '0' <= s[r] <= '9'):
            r -= 1
            continue

        if s[l] != s[r]:
            return False

        l += 1
        r -= 1

    return True


# ------------------ Test (Same as Java main) ------------------

s = "A man, a plan, a canal: Panama"
print(is_palindrome(s))  # Expected: True
